File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3

"""
    This module defines a rectangle.
"""


class Rectangle:
    """
    This class represents a rectangle with the attrubutes.

    width: the width of the rectangle
    height: the height of the rectangle

    It has an instantiation method with optional width and height
    """
    def __init__(self, width=0, height=0):
        """
        Instantiates with optional width and height
        Raises:
        - TypeError: If size is not an integer.
        - ValueError: If size is less than 0.
        """
        if not isinstance(width, int):
            raise TypeError("width must be an integer")
        elif width < 0:
            raise ValueError("width must be >= 0")
        else:
            self.__width = width

        if not isinstance(height, int):
            raise TypeError("height must be an integer")
        elif height < 0:
            raise ValueError("height must be >= 0")
        else:
            self.__height = height

    @property
    def width(self):
        """
        gets the width of the rectangle
        """
        return self.__width

    @width.setter
    def width(self, value):
        """
        sets the width of the rectangle

        Raises:
        - TypeError: If width is not an integer.
        - ValueError: If width is less than 0.
        """
        if not isinstance(value, int):
            raise TypeError("width must be an integer")
        elif value < 0:
            raise ValueError("width must be >= 0")
        else:
            self.__width = value

    @property
    def height(self):
        """
        gets the height of the rectangle
        """
        return self.__height

    @height.setter
    def height(self, value):
        """
        sets the height of the rectangle

        Raises:
        - TypeError: If height is not an integer.
        - ValueError: If height is less than 0.
        """
        if not isinstance(value, int):
            raise TypeError("height must be an integer")
        elif value < 0:
            raise ValueError("height must be >= 0")
        else:
            self.__height = value

    def area(self):
        """
        Returns the area of the rectangle
        """
        return self.__width * self.__height

    def perimeter(self):
        """
        Returns the perimeter of the rectangle
        """
        if self.__width == 0 or self.__height == 0:
            return 0
        else:
            return (self.__width * 2) + (self.__height * 2)

    def __str__(self):
        """
        prints the rectangle in forsm of #
        """
        if self.width == 0 or self.height == 0:
            return ""
        else:
            rectangle_str = ""
            for i in range(self.height):
                for j in range(self.width):
                    rectangle_str += "#"
                rectangle_str += "\n"
            return rectangle_str.rstrip()

    def __repr__(self):
        return f"Rectangle({self.width}, {self.height})"

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

/**
 * class Rectangle that defines a rectangle
 */

class Rectangle {
  /**
   * Rectangle Class
   */
  constructor (w, h) {
    if (w <= 0 || w === undefined || h <= 0 || h === undefined) {
      return this;
    } else {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    /**
     * Instance method
     * prints the rectangle using X
     */

    for (let i = 0; i < this.height; i++) {
      let row = '';
      for (let j = 0; j < this.width; j++) {
        row += 'X';
      }
      console.log(row);
    }
  }

  rotate () {
    /**
     * exchanges the width and the height of the rectangle
     */
    const newHeight = this.height;
    this.height = this.width;
    this.width = newHeight;
  }

  double () {
    /**
     * multiples the width and the height of the rectangle by 2
     */
    this.height *= 2;
    this.width *= 2;
  }
}

module.exports = Rectangle;
